deepinv/training/trainer.py

```diff
@@ -0,0 +1,738 @@
+        """
+        if self.wandb_vis:
+            wandb.log(log_dict_iter)
+
+    def check_clip_grad(self):
+        r"""
+        Check the gradient norm and perform gradient clipping if necessary.
+
+        """
+        out = None
+
+        if self.grad_clip is not None:
+            torch.nn.utils.clip_grad_norm_(self.model.parameters(), self.grad_clip)
+
+        if self.check_grad:
+            # from https://discuss.pytorch.org/t/check-the-norm-of-gradients/27961/7
+            grads = [
+                param.grad.detach().flatten()
+                for param in self.model.parameters()
+                if param.grad is not None
+            ]
+            norm_grads = torch.cat(grads).norm()
+            out = norm_grads.item()
+            self.check_grad_val.update(norm_grads.item())
+
+        return out
+
+    def get_samples_online(self, iterators, g):
+        r"""
+        Get the samples for the online measurements.
+
+        In this setting, a new sample is generated at each iteration by calling the physics operator.
+        This function returns a dictionary containing necessary data for the model inference. It needs to contain
+        the measurement, the ground truth, and the current physics operator, but can also contain additional data.
+
+        :param list iterators: List of dataloader iterators.
+        :param int g: Current dataloader index.
+        :returns: a tuple containing at least: the ground truth, the measurement, and the current physics operator.
+        """
+        data = next(
+            iterators[g]
+        )  # In this case the dataloader outputs also a class label
+
+        if type(data) is tuple or type(data) is list:
+            x = data[0]
+        else:
+            x = data
+
+        x = x.to(self.device)
+
+        physics = self.physics[g]
+
+        y = physics(x)
+
+        return x, y, physics
+
+    def get_samples_offline(self, iterators, g):
+        r"""
+        Get the samples for the offline measurements.
+
+        In this setting, samples have been generated offline and are loaded from the dataloader.
+        This function returns a dictionary containing necessary data for the model inference. It needs to contain
+        the measurement, the ground truth, and the current physics operator, but can also contain additional data.
+
+        :param list iterators: List of dataloader iterators.
+        :param int g: Current dataloader index.
+        :returns: a dictionary containing at least: the ground truth, the measurement, and the current physics operator.
+        """
+        data = next(iterators[g])
+        if (type(data) is not tuple and type(data) is not list) or len(data) != 2:
+            raise ValueError(
+                "If online_measurements=False, the dataloader should output a tuple (x, y)"
+            )
+
+        x, y = data
+        if type(x) is list or type(x) is tuple:
+            x = [s.to(self.device) for s in x]
```

Comment:
Why this is different than how it is in ```get_samples_online``` ?

Reply:
It is for signals x which are TensorLists